Introduce modern unit tests for new test structure (#11525)
PrestaShop/PrestaShop: tests/README.md

Tests
=====

This is PrestaShop automated tests folder.

Multiple type of tests are available:

> - `E2E` and `Selenium` folders contain end-to-end tests maintained by PrestaShop QA team

Reviewer: I don't think it's a good idea to say they are maintained by a specific group. Anyone should be able to add/improve E2E tests, starting with the Core team.

```suggestion
- `E2E` and `Selenium` folders contain end-to-end tests
```

- `Unit` folder contain unit tests, see below

# Unit tests

PrestaShop unit tests are powered by phpunit.

## How to write unit tests

- one php class = one test file
- the test filepath must follow the class filepath

> - every dependency of the class must be mocked*

Reviewer: Watch out, "mock" may not the correct term. According to Martin Fowler, there are actually [several kinds of test doubles](https://martinfowler.com/articles/mocksArentStubs.html#TheDifferenceBetweenMocksAndStubs):

> Meszaros uses the term Test Double as the generic term for any kind of pretend object used in place of a real object for testing purposes. The name comes from the notion of a Stunt Double in movies. (One of his aims was to avoid using any name that was already widely used.) Meszaros then defined four particular kinds of double:
> 
> - **Dummy** objects are passed around but never actually used. Usually they are just used to fill parameter lists.
> - **Fake objects** actually have working implementations, but usually take some shortcut which makes them not suitable for production (an in memory database is a good example).
> - **Stubs** provide canned answers to calls made during the test, usually not responding at all to anything outside what's programmed in for the test.
> - **Spies** are stubs that also record some information based on how they were called. One form of this might be an email service that records how many messages it was sent.
> - **Mocks** are what we are talking about here: objects pre-programmed with expectations which form a specification of the calls they are expected to receive.

In order to avoid misunderstandings, I suggest this change:
```suggestion
- every dependency of the class must be replaced by [test doubles][1]*

[1]: https://martinfowler.com/articles/mocksArentStubs.html#TheDifferenceBetweenMocksAndStubs
```

*If there is a hard-coded dependency such as a singleton pattern being used
or a static call, this class cannot be unit tested and should be tested using
integration tests.

## Conventions

> Use camelCase names for test function names.

Reviewer: ```suggestion
- Use camelCase names for test function names.
- Try to make method names explain the *intent* of the test case as best as possible. Don't hesitate to write long method names if necessary.
	- Bad example: `testGetPrice` (no idea what such a test is supposed to do)
	- Good example: `testDiscountIsAppliedToFinalPrice`
```

Tests
=====

This is PrestaShop automated tests folder.

Multiple type of tests are available:
- `E2E` and `Selenium` folders contain end-to-end tests

> - `Unit` folder contain unit tests, see below

Reviewer 2: containS
